// @ckb-lumos/hd/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AccountExtendedPublicKey", {
  enumerable: true,
  get: function () {
    return _extended_key.AccountExtendedPublicKey;
  }
});
Object.defineProperty(exports, "AddressType", {
  enumerable: true,
  get: function () {
    return _extended_key.AddressType;
  }
});
Object.defineProperty(exports, "ExtendedPrivateKey", {
  enumerable: true,
  get: function () {
    return _extended_key.ExtendedPrivateKey;
  }
});
Object.defineProperty(exports, "ExtendedPublicKey", {
  enumerable: true,
  get: function () {
    return _extended_key.ExtendedPublicKey;
  }
});
Object.defineProperty(exports, "IncorrectPassword", {
  enumerable: true,
  get: function () {
    return _keystore.IncorrectPassword;
  }
});
Object.defineProperty(exports, "InvalidKeystore", {
  enumerable: true,
  get: function () {
    return _keystore.InvalidKeystore;
  }
});
Object.defineProperty(exports, "Keychain", {
  enumerable: true,
  get: function () {
    return _keychain.default;
  }
});
Object.defineProperty(exports, "Keystore", {
  enumerable: true,
  get: function () {
    return _keystore.default;
  }
});
Object.defineProperty(exports, "UnsupportedCipher", {
  enumerable: true,
  get: function () {
    return _keystore.UnsupportedCipher;
  }
});
Object.defineProperty(exports, "XPubStore", {
  enumerable: true,
  get: function () {
    return _xpub_store.XPubStore;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "key", {
  enumerable: true,
  get: function () {
    return _key.default;
  }
});
Object.defineProperty(exports, "mnemonic", {
  enumerable: true,
  get: function () {
    return _mnemonic.default;
  }
});

var _keystore = _interopRequireWildcard(require("./keystore"));

var _xpub_store = require("./xpub_store");

var _keychain = _interopRequireDefault(require("./keychain"));

var _extended_key = require("./extended_key");

var _mnemonic = _interopRequireDefault(require("./mnemonic"));

var _key = _interopRequireDefault(require("./key"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var _default = {
  mnemonic: _mnemonic.default,
  Keystore: _keystore.default,
  UnsupportedCipher: _keystore.UnsupportedCipher,
  IncorrectPassword: _keystore.IncorrectPassword,
  InvalidKeystore: _keystore.InvalidKeystore,
  Keychain: _keychain.default,
  AddressType: _extended_key.AddressType,
  ExtendedPublicKey: _extended_key.ExtendedPublicKey,
  AccountExtendedPublicKey: _extended_key.AccountExtendedPublicKey,
  ExtendedPrivateKey: _extended_key.ExtendedPrivateKey,
  key: _key.default,
  XPubStore: _xpub_store.XPubStore
};
exports.default = _default;
