// @ckb-lumos/hd/lib/keystore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UnsupportedCipher = exports.InvalidKeystore = exports.IncorrectPassword = void 0;

var _crypto = _interopRequireDefault(require("crypto"));

var _sha = require("sha3");

var _uuid = require("uuid");

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _extended_key = require("./extended_key");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class UnsupportedCipher extends Error {
  constructor() {
    super("Unsupported cipher!");
  }

}

exports.UnsupportedCipher = UnsupportedCipher;

class IncorrectPassword extends Error {
  constructor() {
    super("Incorrect password!");
  }

}

exports.IncorrectPassword = IncorrectPassword;

class InvalidKeystore extends Error {
  constructor() {
    super("Invalid keystore, please check your file integrity.");
  }

}

exports.InvalidKeystore = InvalidKeystore;
const CIPHER = "aes-128-ctr";
const CKB_CLI_ORIGIN = "ckb-cli";

// Encrypt and save master extended private key.
class Keystore {
  version = 3;

  constructor(theCrypto, id, origin) {
    this.crypto = theCrypto;
    this.id = id;
    this.origin = origin;
  }

  static fromJson(json) {
    try {
      const object = JSON.parse(json);
      return new Keystore(object.crypto, object.id, object.origin);
    } catch {
      throw new InvalidKeystore();
    }
  }
  /**
   * Load keystore file from path.
   *
   * @param path
   */


  static load(path) {
    const json = _fs.default.readFileSync(path, "utf-8");

    return this.fromJson(json);
  }
  /**
   * Keystore file default name is `${id}.json`.
   *
   * @param dir
   * @param options If you are sure to overwrite existing keystore file, set `overwrite` to true.
   */


  save(dir, {
    name = this.filename(),
    overwrite = false
  } = {}) {
    const path = _path.default.join(dir, name);

    if (!overwrite && _fs.default.existsSync(path)) {
      throw new Error("Keystore file already exists!");
    }

    _fs.default.writeFileSync(path, this.toJson());
  }

  filename() {
    return this.id + ".json";
  }

  toJson() {
    return JSON.stringify(this);
  }

  isFromCkbCli() {
    return this.origin === CKB_CLI_ORIGIN;
  } // Create an empty keystore object that contains empty private key


  static createEmpty() {
    const salt = _crypto.default.randomBytes(32);

    const iv = _crypto.default.randomBytes(16);

    const kdfparams = {
      dklen: 32,
      salt: salt.toString("hex"),
      n: 2 ** 18,
      r: 8,
      p: 1
    };
    return new Keystore({
      ciphertext: "",
      cipherparams: {
        iv: iv.toString("hex")
      },
      cipher: CIPHER,
      kdf: "scrypt",
      kdfparams,
      mac: ""
    }, (0, _uuid.v4)());
  }

  static create(extendedPrivateKey, password, options = {}) {
    const salt = options.salt || _crypto.default.randomBytes(32);

    const iv = options.iv || _crypto.default.randomBytes(16);

    const kdfparams = {
      dklen: 32,
      salt: salt.toString("hex"),
      n: 2 ** 18,
      r: 8,
      p: 1
    };

    const derivedKey = _crypto.default.scryptSync(password, salt, kdfparams.dklen, Keystore.scryptOptions(kdfparams));

    const cipher = _crypto.default.createCipheriv(CIPHER, derivedKey.slice(0, 16), iv);

    if (!cipher) {
      throw new UnsupportedCipher();
    }

    const ciphertext = Buffer.concat([cipher.update(Buffer.from(extendedPrivateKey.serialize().slice(2), "hex")), cipher.final()]);
    return new Keystore({
      ciphertext: ciphertext.toString("hex"),
      cipherparams: {
        iv: iv.toString("hex")
      },
      cipher: CIPHER,
      kdf: "scrypt",
      kdfparams,
      mac: Keystore.mac(derivedKey, ciphertext)
    }, (0, _uuid.v4)());
  } // Imported from xpub with empty private key.


  isEmpty() {
    return this.crypto.ciphertext === "" && this.crypto.mac === "";
  } // Decrypt and return serialized extended private key.


  decrypt(password) {
    const derivedKey = this.derivedKey(password);
    const ciphertext = Buffer.from(this.crypto.ciphertext, "hex");

    if (Keystore.mac(derivedKey, ciphertext) !== this.crypto.mac) {
      throw new IncorrectPassword();
    }

    const decipher = _crypto.default.createDecipheriv(this.crypto.cipher, derivedKey.slice(0, 16), Buffer.from(this.crypto.cipherparams.iv, "hex"));

    return "0x" + Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString("hex");
  }

  extendedPrivateKey(password) {
    return _extended_key.ExtendedPrivateKey.parse(this.decrypt(password));
  }

  checkPassword(password) {
    const derivedKey = this.derivedKey(password);
    const ciphertext = Buffer.from(this.crypto.ciphertext, "hex");
    return Keystore.mac(derivedKey, ciphertext) === this.crypto.mac;
  }

  derivedKey(password) {
    const {
      kdfparams
    } = this.crypto;
    return _crypto.default.scryptSync(password, Buffer.from(kdfparams.salt, "hex"), kdfparams.dklen, Keystore.scryptOptions(kdfparams));
  }

  static mac(derivedKey, ciphertext) {
    return new _sha.Keccak(256).update(Buffer.concat([derivedKey.slice(16, 32), ciphertext])).digest("hex");
  }

  static scryptOptions(kdfparams) {
    return {
      N: kdfparams.n,
      r: kdfparams.r,
      p: kdfparams.p,
      maxmem: 128 * (kdfparams.n + kdfparams.p + 2) * kdfparams.r
    };
  }

}

exports.default = Keystore;
