// @ckb-lumos/common-scripts/lib/p2pkh.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createP2PKHMessageGroup = createP2PKHMessageGroup;

var _base = require("@ckb-lumos/base");

var _helpers = require("@ckb-lumos/helpers");

var _codec = require("@ckb-lumos/codec");

var _bi = require("@ckb-lumos/bi");

function groupInputs(inputs, locks) {
  const lockSet = new Set();

  for (const lock of locks) {
    const scriptHash = _base.utils.ckbHash(_base.blockchain.Script.pack(lock));

    lockSet.add(scriptHash);
  }

  const groups = new Map();

  for (let i = 0; i < inputs.length; i++) {
    const scriptHash = _base.utils.ckbHash(_base.blockchain.Script.pack(inputs[i].cellOutput.lock));

    if (lockSet.has(scriptHash)) {
      if (groups.get(scriptHash) === undefined) groups.set(scriptHash, []);
      groups.get(scriptHash).push(i);
    }
  }

  return groups;
}

function calcRawTxHash(tx) {
  const createdTx = (0, _helpers.createTransactionFromSkeleton)(tx);
  const rawTx = {
    cellDeps: createdTx.cellDeps,
    headerDeps: createdTx.headerDeps,
    inputs: createdTx.inputs,
    outputs: createdTx.outputs,
    outputsData: createdTx.outputsData,
    version: createdTx.version
  };
  return _base.utils.ckbHash(_base.blockchain.RawTransaction.pack(rawTx));
}

const defaultCkbHasher = () => {
  const hasher = new _base.utils.CKBHasher();
  return {
    update: message => hasher.update(message.buffer),
    digest: () => _codec.bytes.bytify(hasher.digestHex())
  };
};

function resolveThunk(thunkOrValue) {
  if (thunkOrValue instanceof Function) return thunkOrValue();
  return thunkOrValue;
}
/**
 * Return an array of messages as well as their corresponding position indexes and locks for signing a P2PKH transaction.
 * For more details, please see:
 * https://github.com/nervosnetwork/ckb-system-scripts/wiki/How-to-sign-transaction
 *
 * @param tx TxSkeleton with all input cells' witnessArgs.lock filled with 0.
 * @param locks Locks you want to sign, e.g. you don't need to sign ACP cells.
 * @param hasher Message hasher, defaults to CKB blake2b hasher. Check
 * https://github.com/nervosnetwork/ckb-system-scripts/blob/e975e8b7d5231fdb1c537b830dd934b305492417/c/secp256k1_blake160_sighash_all.c#L22-L28 for more.
 * @returns An array of Group containing: lock of the input cell you need to sign, message for signing, witness index of this message (first index of the input cell with this lock).
 */


function createP2PKHMessageGroup(tx, locks, {
  hasher: thunkableHasher = defaultCkbHasher
} = {}) {
  const groups = groupInputs(tx.inputs.toArray(), locks);
  const rawTxHash = calcRawTxHash(tx);

  if (locks.length > 1 && !(thunkableHasher instanceof Function)) {
    // If we have multiple locks to group, we need the hasher to be thunk so that in the second group we can get another new hasher.
    throw new Error("Must provide hasher producer when you have multiple locks to group.");
  }

  const messageGroup = [];

  for (const group of groups.keys()) {
    const messageHasher = resolveThunk(thunkableHasher);
    const indexes = groups.get(group);
    const firstIndex = indexes[0];
    const firstWitness = tx.witnesses.get(firstIndex);

    if (firstWitness === undefined) {
      throw new Error("Please fill witnesses with 0 first!");
    }

    messageHasher.update(_codec.bytes.bytify(rawTxHash));
    const lengthBuffer = new ArrayBuffer(8);
    const view = new DataView(lengthBuffer);

    const witnessHexString = _bi.BI.from(_codec.bytes.bytify(firstWitness).length).toString(16);

    if (witnessHexString.length <= 8) {
      view.setUint32(0, Number("0x" + witnessHexString), true);
      view.setUint32(4, Number("0x" + "00000000"), true);
    }

    if (witnessHexString.length > 8 && witnessHexString.length <= 16) {
      view.setUint32(0, Number("0x" + witnessHexString.slice(-8)), true);
      view.setUint32(4, Number("0x" + witnessHexString.slice(0, -8)), true);
    }

    messageHasher.update(new Uint8Array(lengthBuffer));
    messageHasher.update(_codec.bytes.bytify(firstWitness));

    for (let i = 1; i < indexes.length; i++) {
      const witness = tx.witnesses.get(indexes[i]);
      messageHasher.update(new Uint8Array(lengthBuffer));
      messageHasher.update(_codec.bytes.bytify(witness));
    }

    for (let i = tx.inputs.toArray().length; i < tx.witnesses.toArray().length; i++) {
      const witness = tx.witnesses.get(i);
      messageHasher.update(new Uint8Array(lengthBuffer));
      messageHasher.update(_codec.bytes.bytify(witness));
    }

    const digested = messageHasher.digest();
    const g = {
      index: firstIndex,
      lock: tx.inputs.get(firstIndex).cellOutput.lock,
      message: "0x" + Array.prototype.map.call(digested, x => ("00" + x.toString(16)).slice(-2)).join("")
    };
    messageGroup.push(g);
  }

  return messageGroup;
}
