// @ckb-lumos/rpc/lib/types/api.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CKBComponents = void 0;

/**
 * @see https://github.com/nervosnetwork/ckb/blob/develop/protocol/src/protocol.fbs for more infGomation
 */

/* eslint-disable  @typescript-eslint/no-namespace,  @typescript-eslint/no-explicit-any, @typescript-eslint/ban-types  */
let CKBComponents;
exports.CKBComponents = CKBComponents;

(function (_CKBComponents) {
  let TransactionStatus;

  (function (TransactionStatus) {
    TransactionStatus["Pending"] = "pending";
    TransactionStatus["Proposed"] = "proposed";
    TransactionStatus["Committed"] = "committed";
  })(TransactionStatus || (TransactionStatus = {}));

  _CKBComponents.TransactionStatus = TransactionStatus;
  let CellStatus;

  (function (CellStatus) {
    CellStatus["Live"] = "live";
    CellStatus["Unknown"] = "unknown";
  })(CellStatus || (CellStatus = {}));

  _CKBComponents.CellStatus = CellStatus;
})(CKBComponents || (exports.CKBComponents = CKBComponents = {}));
